db/src/types.rs

```diff
@@ -0,0 +1,251 @@
+    pub index: i64,
+}
+
+impl Partition {
+    pub fn new(start: i64, next: i64) -> Partition {
+        assert!(start <= next, "A partition represents a monotonic increasing sequence of entids.");
+        Partition { start: start, index: next }
+    }
+}
+
+/// Map partition names to `Partition` instances.
+pub type PartitionMap = BTreeMap<String, Partition>;
+
+/// A Mentat schema attribute has a value type and several other flags determining how assertions
+/// with the attribute are interpreted.
+///
+/// TODO: consider packing this into a bitfield or similar.
+#[derive(Clone,Debug,Eq,Hash,Ord,PartialOrd,PartialEq)]
+pub struct Attribute {
+    /// The associated value type, i.e., `:db/valueType`?
+    pub value_type: ValueType,
+    /// `true` if this attribute is multi-valued, i.e., it is `:db/cardinality
+    /// :db.cardinality/many`.  `false` if this attribute is single-valued (the default), i.e., it
+    /// is `:db/cardinality :db.cardinality/one`.
+    pub multival: bool,
+    /// `true` if this attribute is unique-value, i.e., it is `:db/unique :db.unique/value`.
+    ///
+    /// *Unique-value* means that there is at most one assertion with the attribute and a
+    /// particular value in the datom store.
+    pub unique_value: bool,
+    /// `true` if this attribute is unique-identity, i.e., it is `:db/unique :db.unique/identity`.
+    ///
+    /// Unique-identity attributes always have value type `Ref`.
+    ///
+    /// *Unique-identity* means that the attribute is *unique-value* and that they can be used in
+    /// lookup-refs and will automatically upsert where appropriate.
+    pub unique_identity: bool,
+    /// `true` if this attribute is automatically indexed, i.e., it is `:db/indexing true`.
+    pub index: bool,
+    /// `true` if this attribute is automatically fulltext indexed, i.e., it is `:db/fulltext true`.
+    ///
+    /// Fulltext attributes always have string values.
+    pub fulltext: bool,
+    /// `true` if this attribute is a component, i.e., it is `:db/isComponent true`.
+    ///
+    /// Component attributes always have value type `Ref`.
+    ///
+    /// They are used to compose entities from component sub-entities: they are fetched recursively
+    /// by pull expressions, and they are automatically recursively deleted where appropriate.
+    pub component: bool,
+}
+
+impl Default for Attribute {
+    fn default() -> Attribute {
+        Attribute {
+            // There's no particular reason to favour one value type, so Ref it is.
+            value_type: ValueType::Ref,
+            fulltext: false,
+            index: false,
+            multival: false,
+            unique_value: false,
+            unique_identity: false,
+            component: false,
+        }
+    }
+}
+
+/// Map `String` idents (`:db/ident`) to positive integer entids (`1`).
+pub type IdentMap = BTreeMap<String, Entid>;
+
+/// Map positive integer entids (`1`) to `String` idents (`:db/ident`).
```

Comment:
I decided to use `BTreeMap` simply because we're using it elsewhere (in `edn`).  Let's handle this in follow-up: https://github.com/mozilla/mentat/issues/192